### PR TITLE
AS-110 Latency Logs

### PR DESCRIPTION
**Pull request by YuriHechter**
Usage example:

archive-backend log latency --output_file="C://latency reports//report.csv"
archive-backend log latency_aggregate --csv_file="C://latency reports//report.csv"

### CODE REVIEW

File: cmd/search_log.go [hunk 1]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
 	}
 	printCsv(records)
 }
+
+func latencyFn(cmd *cobra.Command, args []string) {
+	logger := initLogger()
+	headers := []string{
+		"#", "SearchId", "Term",
+	}
```
Comment (bbfsdev) [inline cmd/search_log.go]: Make this const in this file, then below use len of this slice.

File: cmd/search_log.go [hunk 2]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
...
+		} else {
+			printCsv(records)
+		}
+	}
+	log.Infof("Found %d queries.", totalQueries)
+}
+
+func latencyAggregateFn(cmd *cobra.Command, args []string) {
+
+	const metaColumnsCnt = 3                  //  "#","SearchId","Term"
```
Comment (bbfsdev) [inline cmd/search_log.go]: Len of constant slice (keep the constant in this file).

File: cmd/search_log.go [hunk 3]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
...
+	opLatenciesMap := make(map[string][]int, len(consts.LATENCY_LOG_OPERATIONS_FOR_SEARCH))
+	for _, op := range consts.LATENCY_LOG_OPERATIONS_FOR_SEARCH {
+		opLatenciesMap[op] = make([]int, len(records)-1)
+	}
+
+	for i := 1; i < len(records); i++ { //  skip first line (headers)
+
+		record := records[i]
+
+		for j := metaColumnsCnt; j < len(record); j++ {
```
Comment (bbfsdev) [inline cmd/search_log.go]: again, are the record slice the same size for all queries? What about tomorrow when some stages will be optional?


I suggest to store latencies as slice of pairs  [(stage1, latency1), (stage2, latency2), ...]

File: cmd/search_log.go [hunk 4]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
...
+			lat, err := strconv.Atoi(strings.TrimSpace(record[j]))
+			utils.Must(err)
+			for opIndex, op := range consts.LATENCY_LOG_OPERATIONS_FOR_SEARCH {
+				if opIndex == j-metaColumnsCnt {
+					(opLatenciesMap[op])[i-1] = lat
+					continue
+				}
```
Comment (bbfsdev) [inline cmd/search_log.go]: Not sure that is this index calculation. Can you please explain?

File: utils/general.go [hunk 5]
```diff
@@ -205,3 +205,15 @@ func StringMapOrderedKeys(m interface{}) []string {
 	sort.Strings(keys)
 	return keys
 }
+
+func SumAndMax(values []int) (int, int) {
+	sum := 0
+	max := 0
```
Comment (bbfsdev) [inline utils/general.go]: Initialize max with -max int

File: cmd/search_log.go [hunk 6]
```diff
@@ -58,11 +83,28 @@ func initLogger() *search.SearchLogger {
 	return search.MakeSearchLogger(esManager)
 }
 
+func appendCsvToFile(path string, records [][]string) {
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND, 0660)
+	if err != nil {
+		log.Fatalln("cannot open file: ", err)
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply cmd/search_log.go]: original printCsv func was with 'fatal'

File: consts/consts.go [hunk 7]
```diff
@@ -552,3 +552,28 @@ var GRAMMAR_INTENTS_TO_FILTER_VALUES = map[string]map[string][]string{
...
+// Latency log
+const (
+	LAT_DOSEARCH                          = "DoSearch"
+	LAT_DOSEARCH_MULTISEARCHDO            = "DoSearch.MultisearchDo"
+	LAT_DOSEARCH_MULTISEARCHHIGHLIGHTSDO  = "DoSearch.MultisearcHighlightsDo"
+	LAT_DOSEARCH_ADDINTENTS               = "DoSearch.AddIntents"
+	LAT_DOSEARCH_ADDINTENTS_FIRSTROUNDDO  = "DoSearch.AddIntents.FirstRoundDo"
+	LAT_DOSEARCH_ADDINTENTS_SECONDROUNDDO = "DoSearch.AddIntents.SecondRoundDo"
+	LAT_DOSEARCH_MULTISEARCHTWEETSDO      = "DoSearch.MultisearchTweetsDo"
+	LAT__DOSEARCH_TYPOSUGGESTDO           = "DoSearch.TypoSuggestDo"
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply consts/consts.go]: fixed

File: cmd/search_log.go [hunk 8]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
...
+	if csvFile != "" {
+		appendCsvToFile(csvFile, [][]string{headers})
+	} else {
+		printCsv([][]string{headers})
+	}
+	totalQueries := 0
+	SLICES := 100
+	for i := 0; i < SLICES; i++ {
+		s := elastic.NewSliceQuery().Id(i).Max(SLICES)
+		queries, err := logger.GetLattestQueries(s, null.StringFrom("now-7d/d"), null.BoolFrom(false))
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply cmd/search_log.go]: done

File: cmd/search_log.go [hunk 9]
```diff
@@ -137,3 +179,129 @@ func clicksFn(cmd *cobra.Command, args []string) {
...
+		totalQueries += len(queries)
+		sortedQueries := make(search.CreatedSearchLogs, 0, len(queries))
+		for _, q := range queries {
+			sortedQueries = append(sortedQueries, q)
+		}
+		sort.Sort(sortedQueries)
+		records := [][]string{}
+		for i, sl := range sortedQueries {
+			utils.Must(err)
+			var latencies []string
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply cmd/search_log.go]: preparing this data for CSV writing with io.Writer Write function that expects for string values

[the remark this answers is not in the document]
Comment (YuriHechter) [reply cmd/search_log.go]: The supported latencies are only that is defined in consts.LATENCY_LOG_OPERATIONS_FOR_SEARCH.
I compare OP with each latency from ExecutionTimeLog.
I can explain in more details if this not clear.

File: cmd/search_log.go (hunk 1 above)
Comment (YuriHechter) [reply cmd/search_log.go]: array constant?

File: cmd/search_log.go (hunk 2 above)
Comment (YuriHechter) [reply cmd/search_log.go]: constant slice is not supported in go

File: cmd/search_log.go (hunk 3 above)
Comment (YuriHechter) [reply cmd/search_log.go]: Missing stage values are counted as 0

File: utils/general.go (hunk 5 above)
Comment (YuriHechter) [reply utils/general.go]: done

File: cmd/search_log.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (YuriHechter) [reply cmd/search_log.go]: changed to panic

File: cmd/search_log.go (hunk 2 above)
Comment (YuriHechter) [reply cmd/search_log.go]: changed to len(latencyMetaHeaders)

**Pull request merged**